### PR TITLE
Supports user defined android activities in extensions.

### PR DESCRIPTION
**Pull request by bigeyex**
add @CustomActivities annotation before @SimpleObject.
eg. @CustomActivities(activityNames = "com.google.appinventor.components.runtime.ListPickerActivity")
multiple activities can be separated by commas.

This will add an <activity/> tag in Android Manifest file, giving the app the permission to start the newly defined Activity.
existing classes that involves activities (eg. ListPicker) are unchanged, since they have additional attributes besides just a name in the Manifest file though they works fine with the new CustomActivities annotation.

### CODE REVIEW

File: appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java [hunk 1]
```diff
@@ -231,6 +234,39 @@ void generatePermissions() {
...
+      e.printStackTrace();
+      userErrors.print(String.format(ERROR_IN_STAGE, "CustomActivities"));
+    }
+
+    int n = 0;
+    for (String type : customActivityNeeded.keySet()) {
+      n += customActivityNeeded.get(type).size();
+    }
+
+    System.out.println("Custom Activities needed, n = " + n);
```
Comment (josmas) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java]: please delete

**Comment by jisqyv** [comment]
@bigeyex Can you give us an example where this is necessary. I'm a little unclear on how declaring an activity without an corresponding intent filter helps?

File: appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java [hunk 2]
```diff
@@ -146,6 +147,8 @@
       new ConcurrentHashMap<String, Set<String>>();
   private final ConcurrentMap<String, Set<String>> permissionsNeeded =
       new ConcurrentHashMap<String, Set<String>>();
+  private final ConcurrentMap<String, Set<String>> customActivityNeeded =
```
Comment (josmas) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java]: Maybe call it _customActivitiesNeeded_, in plural

File: appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java [hunk 3]
```diff
@@ -231,6 +234,39 @@ void generatePermissions() {
...
+      n += customActivityNeeded.get(type).size();
+    }
+
+    System.out.println("Custom Activities needed, n = " + n);
+  }
+
+  // Just used for testing
+  @VisibleForTesting
+  Map<String,Set<String>> getCustomActivities() {
+    return permissionsNeeded;
```
Comment (josmas) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java]: this is returning the wrong thing!

File: appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java [hunk 4]
```diff
@@ -504,6 +540,18 @@ private boolean writeAndroidManifest(File manifestFile, Set<String> broadcastRec
         out.write("    </activity>\n");
       }
 
+
+      // Add custom Activities
+      // make custom activities unique by putting them in one set
+      Set<String> customActivities = Sets.newHashSet();
```
Comment (josmas) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java]: this dedup should probably happen before getting to this point; maybe in the method that creates _customActivitiesNeeded_

File: appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java [hunk 5]
```diff
@@ -89,6 +89,7 @@
   public static final String NATIVE_TARGET = "native";
   // Must match ComponentListGenerator.PERMISSIONS_TARGET
   private static final String PERMISSIONS_TARGET = "permissions";
+  private static final String CUSTOMACTIVITIES_TARGET = "customActivities";
```
Comment (josmas) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java]: _CUSTOM_ACTIVITIES_TARGET_ may be easier to read

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentListGenerator.java [hunk 6]
```diff
@@ -22,6 +22,7 @@
 public final class ComponentListGenerator extends ComponentProcessor {
   // Names of component information types to be output. Must match buildserver.compiler constants.
   private static final String PERMISSIONS_TARGET = "permissions";
+  private static final String CUSTOMACTIVITIES_TARGET = "customActivities";
```
Comment (josmas) [inline appinventor/components/src/com/google/appinventor/components/scripts/ComponentListGenerator.java]: same as above

**Comment by josmas** [comment]
@bigeyex thanks for this!
I left a few inline comments. My main problem with this is that if the _ListPicker_ needs more attributes, they should be added to the annotation.

A change like this can only be easily tested with an examples of its use, so _ListPicker_ would be perfect.

**Comment by josmas** [comment]
@jisqyv more context in this thread: https://groups.google.com/forum/#!topic/app-inventor-open-source-dev/MDA09K33W20

**Comment by jisqyv** [comment]
@bigeyex @josmas I see the value of this change. But wonder if we can make it a bit more general, to add other information to the manifest as well.

**Comment by josmas** [comment]
what kind of extra info @jisqyv ? you mean more info about the custom activity? I would certainly agree with that.

If it's not usable for existing components like _ListPicker_, _WebView_, _BarcodeScanner_, and _Twitter_, which use an extra activities, then it's of less value than it should.

**Comment by bigeyex** [comment]
@josmas @jisqyv , I checked up those existing components. They are more than just using attributes in Activity declaration - they use intent-filters. Should I allow users to write the complete activity tag (<activity android:name="..." android:...><intent-filters.....></activity>) in their annotation?

**Comment by ewpatton** [comment]
The behavior of this PR is subsumed by the changes William Byrne did to add annotations for modifying the AndroidManifest.xml at compile time.

**Pull request closed without merging**